### DISCUSSION TITLE
Using aux variables and material properties in DerivativeParsedMaterial

### DISCUSSION THREAD

**Discussion by rnpratoori** [idaholab/moose#28876] [Q&A Modules: Phase field]
Hello,

I am working on a set of phase-field equations where the energy is defined using polymer volume fraction instead of the regular phase field order parameter `c`. Although I can rewrite the expression using `c`, it would make it cumbersome.

Now, to combat this, I have defined an aux variable for polymer volume fraction which is calculated based on `c`. Can I use this to define the expression for energy using `DerivativeParsedMaterial`? In this case, what will be the `coupled_variable`? 'c' or polymer volume fraction?

I also have a material property which is calculated at each quadrature point and used in one of the energy terms and is also dependent on polymer volume fraction. How is it handled in `DerivativeParsedMaterial`?

**Comment by GiudGiud** [comment]
Hello

An auxiliary variable can be substituted but I would not recommend it if you are using automatic differentiation (AD) or the Derivatives from the DerivativeParsedMaterial because auxiliary variables do not propagate derivatives (there might be a way with the DerivativeParsedMaterial, definitely not with AD)

So instead I would recommend using an AD material property instead

**Reply by rnpratoori** [reply]
So, having defining both as the material property with `DerivativeParsedMaterial` and then using them in another material property will work? In this case I believe I need to use `c` as the `coupled_variable`.

**Reply by GiudGiud** [reply]
I think so. I recommend you check with the Jacobian debugger afterwards
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

**Reply by rnpratoori** [reply]
So, I tried with `u` (phase field order parameter) and I am getting this error - 
```
*** ERROR ***
Invalid function
(-1/(N2*phi^2))*(log(1-phi)+phi+nc*vs*((1/phi)-(phi/2)))
u
in ParsedMaterialHelper.
Syntax error: Unknown identifier
```
I tried with `phi` (polymer volume fraction) and it throws this error -
```
*** ERROR ***
chi: Coupled variable 'phi' was not found
```
I am not sure what I am missing here...
This is the complete input file - https://github.com/rnpratoori/panther/blob/ddcfc03e83ae240f7632eb57a575d3d433dc6981/problems/CL_CHNS.i

**Reply by GiudGiud** [reply]
```
    [./chi]
        type = DerivativeParsedMaterial
        property_name = chi
        coupled_variables = 'u'
        constant_names =        'N2     nc      vs      s'
        constant_expressions = '${N2}   ${nc}   ${vs}   ${s}'
        expression = '(-1/(N2*phi^2))*(log(1-phi)+phi+nc*vs*((1/phi)-(phi/2)))'
        derivative_order = 2
    [../]
```
should likely be
```
    [./chi]
        type = DerivativeParsedMaterial
        property_name = chi
        coupled_variables = 'u phi'
        constant_names =        'N2     nc      vs      s'
        constant_expressions = '${N2}   ${nc}   ${vs}   ${s}'
        expression = '(-1/(N2*phi^2))*(log(1-phi)+phi+nc*vs*((1/phi)-(phi/2)))'
        derivative_order = 2
    [../]
```
if phi is a variable

**Reply by rnpratoori** [reply]
`phi` here is another `DerivativeParsedMaterial` which is dependent on the variable `u`. How do I handle it in this case?

**Reply by GiudGiud** [reply]
add to the Material
```
material_property_names = phi
```

**Reply by rnpratoori** [reply]
I was just looking at the options for `DerivativeParsedMaterial`, should `phi` not be a `additional_derivative_symbols`

**Reply by GiudGiud** [reply]
Unlikely, we only need variables there. When assembling a Jacobian we only need derivatives wrt to variables. 

If we needed the derivative of this property with regards to phi, it would be for another object s needs.

**Reply by rnpratoori** [reply]
I wanted to check the Jacobian with the debugger as you suggested. It gives an error message - 
```
 /home/rnp/Software/MOOSE/projects/moose/python/jacobiandebug/analyzejacobian.py CL_CHNS.i
Traceback (most recent call last):
  File "/home/rnp/Software/MOOSE/projects/moose/python/jacobiandebug/analyzejacobian.py", line 339, in <module>
    executable = findExecutable(options.executable, options.method)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rnp/Software/MOOSE/projects/moose/python/jacobiandebug/analyzejacobian.py", line 82, in findExecutable
    elif shutil.which(executable_option):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rnp/Software/miniforge/envs/moose/lib/python3.11/shutil.py", line 1493, in which
    if os.path.dirname(cmd):
       ^^^^^^^^^^^^^^^^^^^^
  File "<frozen posixpath>", line 152, in dirname
TypeError: expected str, bytes or os.PathLike object, not NoneType
```
Is there something wrong I am doing?
I used this command to run the debugger -
```
/home/rnp/Software/MOOSE/projects/moose/python/jacobiandebug/analyzejacobian.py CL_CHNS.i
```

**Reply by GiudGiud** [reply]
you need to provide the executable if it can't find it
```
/home/rnp/Software/MOOSE/projects/moose/python/jacobiandebug/analyzejacobian.py -e <path to your executable> CL_CHNS.i
```
